src/dune_rules/lib_rules.ml

```diff
@@ -487,6 +487,27 @@ let setup_build_archives (lib : Library.t) ~top_sorted_modules ~cctx ~expander ~
     (fun () -> build_shared ~native_archives ~sctx lib ~dir ~flags)
 ;;
 
+let parameterize_root_module ~parameter ~root_module_name modules =
+  let* parameter_module_name =
+    let+ name = Lib.main_module_name parameter in
+    match Resolve.to_result name with
+    | Ok (Some name) -> name
+    | Ok None ->
+      (* The parameter must have a module name otherwise it can't exist. *)
+      assert false
+    | Error err -> Resolve.raise_error_with_stack_trace err
```

Comment:
Per https://github.com/ocaml/dune/pull/12002#discussion_r2248845061

> This error needs to be delayed rather than forced.

Reply:
Indeed! The error has been delayed until we really need it.